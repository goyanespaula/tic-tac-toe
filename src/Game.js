import React, { Component } from "react";
import Board from "./containers/Board";

class Game extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currPlayer: "X",
      showWinScreen: false,
      layout: Array(3)
        .fill(null)
        .map(square => Array(3).fill(null))
    };
    this.playSquare = this.playSquare.bind(this);
  }

  changePlayer() {
    let currPlayer = this.state.currPlayer === "X" ? "O" : "X";
    this.setState({ currPlayer });
  }

  playSquare(x, y) {
    // if (checkWin())
    // else
    let newLayout = this.state.layout.slice();
    newLayout[x][y] = this.state.currPlayer === "X" ? "X" : "O";
    this.changePlayer();
    this.setState(newLayout);
  }

  reset() {
    this.setState({ showWinScreen: false, currPlayer: 1 });
  }

  render() {
    return (
      <div
        style={{
          display: "flex",
          flexDirection: "column",
          justifyContent: "center",
          alignItems: "center"
        }}
      >
        <header>
          <h1>Welcome to TicTacToe</h1>
        </header>
        <Board layout={this.state.layout} playSquare={this.playSquare} />
      </div>
    );
  }
}

export default Game;
