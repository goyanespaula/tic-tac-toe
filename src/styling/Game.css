#game {
  margin-top: 2vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

header {
  text-align: center;
  color: white;
  font-size: 1.25rem;
}

h2 {
  margin: 2rem;
  color: rgba(255, 255, 255, 0.9);
}
