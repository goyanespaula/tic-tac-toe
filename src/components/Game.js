import React, { Component } from "react";
import Board from "./Board";
import WinScreen from "./WinScreen";
import checkWin from "../helpers/checkWin";
import "../styling/Game.css";

const initialState = {
  currPlayer: "X",
  layout: Array(3)
    .fill(null)
    .map(square => Array(3).fill(null)),
  moveCount: 0,
  winner: false,
  showWinScreen: false
};

class Game extends Component {
  constructor(props) {
    super(props);
    this.state = initialState;
    this.playSquare = this.playSquare.bind(this);
    this.reset = this.reset.bind(this);
  }

  changePlayer() {
    let currPlayer = this.state.currPlayer === "X" ? "O" : "X";
    this.setState({ currPlayer });
  }

  playSquare(y, x) {
    let layout = this.state.layout.map(arr => arr.slice());
    layout[y][x] = this.state.currPlayer;
    let moveCount = this.state.moveCount + 1;

    const winner = checkWin(layout, y, x, this.state.currPlayer, moveCount);

    this.setState({ layout, moveCount });

    if (winner) {
      this.setState({ winner, showWinScreen: true });
    } else {
      this.changePlayer();
    }
  }

  reset() {
    this.setState(initialState);
  }

  render() {
    return (
      <div id="game">
        <header>
          <h1>Tic-Tac-Toe</h1>
        </header>
        <Board layout={this.state.layout} playSquare={this.playSquare} />
        <h2>Current Player: {this.state.currPlayer}</h2>
        {this.state.showWinScreen ? (
          <WinScreen reset={this.reset} winner={this.state.winner} />
        ) : null}
      </div>
    );
  }
}

export default Game;
