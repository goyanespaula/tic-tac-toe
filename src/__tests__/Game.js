import React from "react";
import { shallow } from "enzyme";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import Game from "../Game";

describe("<Game /> initial state", () => {
  let wrapper;
  it("renders", () => {
    wrapper = shallow(<Game />);
  });

  it("matches snapshot", () => {
    const serialized = toJson(wrapper);
    expect(serialized).toMatchSnapshot();
  });

  it("should have 'X' as initial player", () => {
    wrapper = mount(<Game />);
    expect(wrapper.state("currPlayer")).toBe("X");
  });

  it("should have an initial layout of null array of arrays", () => {
    expect(wrapper.state("layout")).toEqual([
      [null, null, null],
      [null, null, null],
      [null, null, null]
    ]);
  });

  it("should have an initial move count is 0", () => {
    expect(wrapper.state("moveCount")).toBe(0);
  });

  it("should have no initial winner", () => {
    expect(wrapper.state("winner")).toBe(false);
  });

  it("should have win screen initially frozen", () => {
    expect(wrapper.state("showWinScreen")).toBe(false);
  });
});

describe("<Game /> while played", () => {
  const wrapper = mount(<Game />);
  it("should update currPlayer, layout, and moveCount when square is clicked", () => {
    wrapper
      .find("[x=0]")
      .find("[y=0]")
      .simulate("click");
    expect(wrapper.state()).toEqual({
      currPlayer: "O",
      layout: [["X", null, null], [null, null, null], [null, null, null]],
      moveCount: 1,
      winner: false,
      showWinScreen: false
    });
  });

  it("should not change state when square is double clicked", () => {
    wrapper
      .find("[y=0]")
      .find("[x=0]")
      .simulate("click");
    expect(wrapper.state()).toEqual({
      currPlayer: "O",
      layout: [["X", null, null], [null, null, null], [null, null, null]],
      moveCount: 1,
      winner: false,
      showWinScreen: false
    });
  });

  it("should update state, including winner and showWinScreen when game is won by 'X'", () => {
    wrapper
      .find("[y=1]")
      .find("[x=0]")
      .simulate("click");
    wrapper
      .find("[y=1]")
      .find("[x=1]")
      .simulate("click");
    wrapper
      .find("[y=0]")
      .find("[x=1]")
      .simulate("click");
    wrapper
      .find("[y=2]")
      .find("[x=2]")
      .simulate("click");
    expect(wrapper.state()).toEqual({
      currPlayer: "X",
      layout: [["X", "O", null], ["O", "X", null], [null, null, "X"]],
      moveCount: 5,
      winner: "X",
      showWinScreen: true
    });
  });

  it("should reset state when reset button is clicked", () => {
    wrapper.find("#reset-btn").simulate("click");
    expect(wrapper.state()).toEqual({
      currPlayer: "X",
      layout: [[null, null, null], [null, null, null], [null, null, null]],
      moveCount: 0,
      winner: false,
      showWinScreen: false
    });
  });

  it("should update state, including winner and showWinScreen when game is won by 'O'", () => {
    wrapper
      .find("[y=1]")
      .find("[x=2]")
      .simulate("click");
    wrapper
      .find("[y=1]")
      .find("[x=1]")
      .simulate("click");
    wrapper
      .find("[y=1]")
      .find("[x=0]")
      .simulate("click");
    wrapper
      .find("[y=2]")
      .find("[x=2]")
      .simulate("click");
    wrapper
      .find("[y=0]")
      .find("[x=0]")
      .simulate("click");
    wrapper
      .find("[y=2]")
      .find("[x=0]")
      .simulate("click");
    wrapper
      .find("[y=0]")
      .find("[x=2]")
      .simulate("click");
    wrapper
      .find("[y=2]")
      .find("[x=1]")
      .simulate("click");
    expect(wrapper.state()).toEqual({
      currPlayer: "O",
      layout: [["X", null, "X"], ["X", "O", "X"], ["O", "O", "O"]],
      moveCount: 8,
      winner: "O",
      showWinScreen: true
    });
  });

  it("should update state when game is a draw", () => {
    wrapper.find("#reset-btn").simulate("click");
    wrapper
      .find("[y=1]")
      .find("[x=2]")
      .simulate("click");
    wrapper
      .find("[y=1]")
      .find("[x=1]")
      .simulate("click");
    wrapper
      .find("[y=1]")
      .find("[x=0]")
      .simulate("click");
    wrapper
      .find("[y=2]")
      .find("[x=2]")
      .simulate("click");
    wrapper
      .find("[y=0]")
      .find("[x=0]")
      .simulate("click");
    wrapper
      .find("[y=2]")
      .find("[x=0]")
      .simulate("click");
    wrapper
      .find("[y=0]")
      .find("[x=2]")
      .simulate("click");
    wrapper
      .find("[y=0]")
      .find("[x=1]")
      .simulate("click");
    wrapper
      .find("[y=2]")
      .find("[x=1]")
      .simulate("click");
    expect(wrapper.state()).toEqual({
      currPlayer: "X",
      layout: [["X", "O", "X"], ["X", "O", "X"], ["O", "X", "O"]],
      moveCount: 9,
      winner: "draw",
      showWinScreen: true
    });
  });
});
