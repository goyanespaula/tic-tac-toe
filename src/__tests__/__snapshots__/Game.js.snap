// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> initial state matches snapshot 1`] = `
<div
  id="game"
>
  <header>
    <h1>
      Tic-Tac-Toe
    </h1>
  </header>
  <Board
    layout={
      Array [
        Array [
          null,
          null,
          null,
        ],
        Array [
          null,
          null,
          null,
        ],
        Array [
          null,
          null,
          null,
        ],
      ]
    }
    playSquare={[Function]}
  />
  <h2>
    Current Player: 
    X
  </h2>
</div>
`;
